src/gsy_e/models/strategy/energy_parameters/heatpump/heat_pump.py

```diff
@@ -76,6 +121,108 @@ def tanks(self) -> AllTanksEnergyParameters:
         """Exposes the tanks state."""
         return self._tanks_state
 
+    def get_energy_to_buy_maximum_kWh(self, time_slot: DateTime, source_temp_C: float) -> float:
+        """Get maximum energy to buy from the heat pump + storage."""
+        max_heat_demand_kW = (
+            convert_kWh_to_W(
+                self._charger.get_max_heat_energy_charge_kWh(time_slot), self._slot_length
+            )
+            / 1000.0
+        )
+        cop = self._cop_model.calc_cop(
+            source_temp_C=source_temp_C,
+            condenser_temp_C=self._charger.get_condenser_temperature_C(time_slot),
+            heat_demand_kW=max_heat_demand_kW,
+        )
+
+        if cop == 0:
+            return 0
+
+        max_energy_consumption_kWh = max_heat_demand_kW / cop
```

Comment:
The COP does not convert between kW and kWh. It only converts between heat and electricity energy (or power). 
This is a bug AFAIU.

Reply:
Indeed and this should be removed by now.